### ISSUE TITLE
unable to 'make test'

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2301]
Migrated from [rt.perl.org#3625](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3625) (status was 'resolved')

Searchable as RT3625$

**Comment by p5pRT** [comment]
### From mstevens@trinity.fluff.org

#### Created by mstevens@trinity.fluff.org



Unable to run test suite&#8203;:

\-\-cut\-\-
mstevens@&#8203;trinity&#8203;:\~/bleadperl> make test
make&#8203;: \*\*\* No rule to make target \`\.patch'\, needed by \`patchlevel\.h'\.  Stop\.
\-\-cut\-\-

Today's bleadperl\, built with 'sh Configure \-de; make;'\.

<details><summary>Perl Info</summary>

```

Flags:
    category=install
    severity=medium

Site configuration information for perl v5.6.0:

Configured by mstevens at Thu Aug  3 04:52:10 BST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.14, archname=i586-linux
    uname='linux trinity 2.2.14 #2 fri apr 14 02:17:55 bst 2000 i586 unknown '
    config_args='-de'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2', gccversion=2.7.2.3, gccosandvers=
    cppflags='-I/usr/local/include'
    ccflags ='-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldbm -ldb -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.0.7.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    devel-6502


@INC for perl v5.6.0:
    lib
    /usr/local/lib/perl5/5.6.0/i586-linux
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/i586-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/mstevens
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/bin:/bin:/usr/bin:/usr/games:/sbin:/usr/sbin:/usr/local/sbin:/usr/local/bin:/home/mstevens/bin
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Michael Stevens \(lists\.p5p\)&#8203;:
>mstevens@&#8203;trinity&#8203;:\~/bleadperl> make test
>make&#8203;: \*\*\* No rule to make target \`\.patch'\, needed by \`patchlevel\.h'\.  Stop\.

This is my idiocy\. fix\_pl deliberately destroys \.patch to stop
patchlevel\.h being fixed up every time make is run\. This will fix it&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- Makefile.SH~        Thu Aug  3 13:12:41 2000
+++ Makefile.SH Thu Aug  3 13:15:12 2000
@@ -328,6 +328,7 @@
 if test -f .patch ; then $spitshell >>Makefile <<'!NO!SUBS!'
 patchlevel.h: .patch
 	perl fix_pl || (make -f Makefile.micro && ./microperl fix_pl)
+	$(SHELL) Makefile.SH

 !NO!SUBS!

But there is probably a cleaner way.
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Thu\, Aug 03\, 2000 at 04&#8203;:17&#8203;:00AM \+0000\, Simon Cozens wrote&#8203;:
> Michael Stevens \(lists\.p5p\)&#8203;:
> >mstevens@&#8203;trinity&#8203;:\~/bleadperl> make test
> >make&#8203;: \*\*\* No rule to make target \`\.patch'\, needed by \`patchlevel\.h'\.  Stop\.
> This is my idiocy\. fix\_pl deliberately destroys \.patch to stop
> patchlevel\.h being fixed up every time make is run\. This will fix it&#8203;:
> \-\-\- Makefile\.SH~        Thu Aug  3 13&#8203;:12&#8203;:41 2000
> \+\+\+ Makefile\.SH Thu Aug  3 13&#8203;:15&#8203;:12 2000
> @&#8203;@&#8203; \-328\,6 \+328\,7 @&#8203;@&#8203;
>  if test \-f \.patch ; then $spitshell >>Makefile \<\<'\!NO\!SUBS\!'
>  patchlevel\.h&#8203;: \.patch
>      perl fix\_pl || \(make \-f Makefile\.micro && \./microperl fix\_pl\)
> \+    $\(SHELL\) Makefile\.SH
>  \!NO\!SUBS\!
> But there is probably a cleaner way\.

This seems to work\.

**Comment by p5pRT** [comment]
### From @jhi

On Thu\, Aug 03\, 2000 at 05&#8203;:37&#8203;:59AM \+0100\, Michael Stevens wrote&#8203;:
> On Thu\, Aug 03\, 2000 at 04&#8203;:17&#8203;:00AM \+0000\, Simon Cozens wrote&#8203;:
> > Michael Stevens \(lists\.p5p\)&#8203;:
> > >mstevens@&#8203;trinity&#8203;:\~/bleadperl> make test
> > >make&#8203;: \*\*\* No rule to make target \`\.patch'\, needed by \`patchlevel\.h'\.  Stop\.
> > This is my idiocy\. fix\_pl deliberately destroys \.patch to stop
> > patchlevel\.h being fixed up every time make is run\. This will fix it&#8203;:
> > \-\-\- Makefile\.SH~        Thu Aug  3 13&#8203;:12&#8203;:41 2000
> > \+\+\+ Makefile\.SH Thu Aug  3 13&#8203;:15&#8203;:12 2000
> > @&#8203;@&#8203; \-328\,6 \+328\,7 @&#8203;@&#8203;
> >  if test \-f \.patch ; then $spitshell >>Makefile \<\<'\!NO\!SUBS\!'
> >  patchlevel\.h&#8203;: \.patch
> >      perl fix\_pl || \(make \-f Makefile\.micro && \./microperl fix\_pl\)
> > \+    $\(SHELL\) Makefile\.SH
> >  \!NO\!SUBS\!
> > But there is probably a cleaner way\.
> 
> This seems to work\.

Thusly\, it's in\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

simon@&#8203;brecon\.co\.uk \(Simon Cozens\) wrote
> This is my idiocy\. fix\_pl deliberately destroys \.patch to stop
> patchlevel\.h being fixed up every time make is run\. This will fix it&#8203;:

Err\.\.\. no it won't\.    You obviously only tested it once\.   :\-\)

It falls flat on its face the \*second\* time you run make\.

You need to test the existence of \.patch when make is run\, not when
Configure \(and therefore Makefile\.SH\) is run\.

But what are these antics for anyway?\.   Surely the normal "make"
mechanisms will ensure that the code is only run if patchlevel\.h is
older than \.patch ?

Also note that almost everything has patchlevel\.h as a dependency\.
So this will cause \*everything\* to be rebuilt whenever the patch
number is bumped\.    I suffered from this with my local patch\-applying
system\, which dutifully puts entries in patchlevel\.h\.    I ducked the
problem by making my patch\-applying system reset the timestamp\.   Blech\.


Mike Guy

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Mike Guy \(lists\.p5p\)&#8203;:
>simon@&#8203;brecon\.co\.uk \(Simon Cozens\) wrote
>> This is my idiocy\. fix\_pl deliberately destroys \.patch to stop
>> patchlevel\.h being fixed up every time make is run\. This will fix it&#8203;:
>
>Err\.\.\. no it won't\.    You obviously only tested it once\.   :\-\)
>
>It falls flat on its face the \*second\* time you run make\.

Yes\, I know\.
This is precisely \*because\* fix\_pl deliberately destroys \.patch
Which is what I said\.

>But what are these antics for anyway?\.   Surely the normal "make"
>mechanisms will ensure that the code is only run if patchlevel\.h is
>older than \.patch ?

You'd think\.\.\. :/

>Also note that almost everything has patchlevel\.h as a dependency\.
>So this will cause \*everything\* to be rebuilt whenever the patch
>number is bumped\.

I know\. I said there had to be a better way\. Since people are going to
be updating via rsync and then running make to perpetrate the changes\,
it has to be somewhere in the Makefile\. If you can provide a better way
of doing it\, I'm all ears\.

**Comment by p5pRT** [comment]
### From @jhi

On Thu\, Aug 03\, 2000 at 02&#8203;:47&#8203;:29PM \+0000\, Simon Cozens wrote&#8203;:
> Mike Guy \(lists\.p5p\)&#8203;:
> >simon@&#8203;brecon\.co\.uk \(Simon Cozens\) wrote
> >> This is my idiocy\. fix\_pl deliberately destroys \.patch to stop
> >> patchlevel\.h being fixed up every time make is run\. This will fix it&#8203;:
> >
> >Err\.\.\. no it won't\.    You obviously only tested it once\.   :\-\)
> >
> >It falls flat on its face the \*second\* time you run make\.
> 
> Yes\, I know\.
> This is precisely \*because\* fix\_pl deliberately destroys \.patch
> Which is what I said\.

Argle\.\.\. I disabled the Makefile\.SH fix\_pl rule for now until
someone comes up with a robust scheme for it\.

**Comment by p5pRT** [comment]
### From @ysth

In article \<20000803105103\.A2384@&#8203;chaos\.wustl\.edu>\,
Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> wrote&#8203;:
> On Thu\, Aug 03\, 2000 at 02&#8203;:47&#8203;:29PM \+0000\, Simon Cozens wrote&#8203;:
> > Mike Guy \(lists\.p5p\)&#8203;:
> > >simon@&#8203;brecon\.co\.uk \(Simon Cozens\) wrote
> > >> This is my idiocy\. fix\_pl deliberately destroys \.patch to stop
> > >> patchlevel\.h being fixed up every time make is run\. This will fix it&#8203;:
> > >
> > >Err\.\.\. no it won't\.    You obviously only tested it once\.   :\-\)
> > >
> > >It falls flat on its face the \*second\* time you run make\.
> > 
> > Yes\, I know\.
> > This is precisely \*because\* fix\_pl deliberately destroys \.patch
> > Which is what I said\.
> 
> Argle\.\.\. I disabled the Makefile\.SH fix\_pl rule for now until
> someone comes up with a robust scheme for it\. 

Is there something wrong with just requiring the builder to use
\-DPATCHLEVEL=6500 on the make?  Why muck about with a \.patch file?

Anyway\, fix\_pl fails on os2\, which doesn't rename over existing files\.
\(Also\, C\<print FH;> trips my ambiguousity sensors\-\-but I know thats just
a personal thing\)\.

<details><summary>Inline Patch</summary>

```diff
--- fix_pl.orig	Thu Aug  3 00:09:52 2000
+++ fix_pl	Thu Aug  3 13:23:24 2000
@@ -1,7 +1,9 @@
 #!perl
 # Not fixing perl, but fixing the patchlevel if this perl comes
 # from the repository rather than an official release
-exit unless -e ".patch";
+BEGIN {exit unless -e ".patch";}
+use File::Copy 'move';
+
 open PATCH, ".patch" or die "Couldn't open .patch: $!";
 open PLIN, "patchlevel.h" or die "Couldn't open patchlevel.h : $!";
 open PLOUT, ">patchlevel.new" or die "Couldn't write on patchlevel.new : $!";
@@ -14,8 +16,9 @@
         print PLOUT "\t,\"devel-$pl\"\n";
     }
     $seen++ if /local_patches\[\]/;
-    print PLOUT;
+    print PLOUT $_;
 }
-close PLOUT; close PLIN;
-rename "patchlevel.new", "patchlevel.h" or die "Couldn't rename: $!";
+close PLOUT or die "Couldn't close patchlevel.new: $!";
+close PLIN or die "Couldn't close patchlevel.h: $!";
+move("patchlevel.new", "patchlevel.h") or die "Couldn't rename: $!";
 unlink ".patch";

```
</details>
End of Patch.

If you think relying on File&#8203;::Copy of the existing perl installation
is a bad idea\, perhaps a C\<\#\!perl \-pi patchlevel\.h> approach could be
taken instead\.
